piet/src/render_context.rs

```diff
@@ -201,3 +204,86 @@ impl<P: RenderContext> IntoBrush<P> for Color {
         Cow::Owned(piet.solid_brush(self.to_owned()))
     }
 }
+
+/// A color or a gradient.
+///
+/// This type is provided as a convenience, so that library consumers can
+/// easily write methods and types that use or reference *something* that can
+/// be used as a brush, without needing to know what it is.
+///
+/// # Examples
+///
+/// ```no_run
+/// use piet::{Color, PaintBrush, RadialGradient};
+/// use piet::kurbo::Rect;
+///
+/// struct Widget {
+/// frame: Rect,
+/// background: PaintBrush,
+/// }
+///
+/// fn make_widget<T: Into<PaintBrush>>(frame: Rect, bg: T) -> Widget {
+///     Widget {
+///         frame,
+///         background: bg.into(),
+///     }
+/// }
+///
+/// let color_widget = make_widget(Rect::ZERO, Color::BLACK);
+/// let rad_grad = RadialGradient::new(0.8, (Color::WHITE, Color::BLACK));
+/// let gradient_widget = make_widget(Rect::ZERO, rad_grad);
+///
+/// ```
+pub enum PaintBrush {
+    Color(Color),
+    Linear(LinearGradient),
+    Radial(RadialGradient),
+    Fixed(FixedGradient),
+}
+
+impl<P: RenderContext> IntoBrush<P> for PaintBrush {
+    fn make_brush<'a>(&'a self, piet: &mut P, bbox: impl FnOnce() -> Rect) -> Cow<'a, P::Brush> {
+        match self {
+            PaintBrush::Color(color) => color.make_brush(piet, bbox),
+            PaintBrush::Linear(linear) => linear.make_brush(piet, bbox),
+            PaintBrush::Radial(radial) => radial.make_brush(piet, bbox),
+            PaintBrush::Fixed(fixed) => fixed.make_brush(piet, bbox),
+        }
+    }
+}
+
+impl From<Color> for PaintBrush {
+    fn from(src: Color) -> PaintBrush {
+        PaintBrush::Color(src)
+    }
+}
+
+impl From<LinearGradient> for PaintBrush {
+    fn from(src: LinearGradient) -> PaintBrush {
+        PaintBrush::Linear(src)
+    }
+}
+
+impl From<RadialGradient> for PaintBrush {
+    fn from(src: RadialGradient) -> PaintBrush {
+        PaintBrush::Radial(src)
+    }
+}
+
+impl From<FixedGradient> for PaintBrush {
+    fn from(src: FixedGradient) -> PaintBrush {
+        PaintBrush::Fixed(src)
+    }
+}
+
+impl From<FixedLinearGradient> for PaintBrush {
+    fn from(src: FixedLinearGradient) -> PaintBrush {
+        PaintBrush::Fixed(src.into())
```

Comment:
A curiosity: this can probably be written `src.into().into()`. But writing it out more explicitly is probably better.

Reply:
my hunch is no, double inference falls over even in places where we'd expect it to be totally clear.

(tested out of curiosity, and yea this doesn't work 😢)